src/io/udp_frame.rs

```diff
@@ -0,0 +1,251 @@
+    rd: Vec<u8>,
+    wr: Vec<u8>,
+}
+
+impl<C : CodecUdp> Stream for FramedUdp<C> {
+    type Item = C::In;
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Option<C::In>, io::Error> {
+        loop {
+
+            let ret = self.socket.recv_from(self.rd.as_mut_slice());
+            match ret {
+                Ok((n, addr)) => { 
+                    trace!("read {} bytes", n);
+                    trace!("attempting to decode a frame");
+                    if let Some(frame) = try!(self.codec.decode(&addr, & self.rd[.. n])) {
+                        trace!("frame decoded from buffer");
+                        return Ok(Async::Ready(Some(frame)));
+                    }
+                }
+                Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+                    return Ok(Async::NotReady)
+                }
+                Err(e) => return Err(e),
+            }
+        }
+    }
+}
+
+impl<C : CodecUdp> Sink for FramedUdp<C> {
+    type SinkItem = C::Out;
+    type SinkError = io::Error;
+
+    fn start_send(&mut self, item: C::Out) -> StartSend<C::Out, io::Error> {
+        if self.wr.len() > 0 {
+            try!(self.poll_complete());
+            if self.wr.len() > 0 {
+                return Ok(AsyncSink::NotReady(item));
+            }
+        }
+
+        self.out_addr = Some(self.codec.encode(&item, &mut self.wr));
+        Ok(AsyncSink::Ready)
+    }
+
+    fn poll_complete(&mut self) -> Poll<(), io::Error> {
+        trace!("flushing framed transport");
+
+        while !self.wr.is_empty() {
+            if let Some(outaddr) = self.out_addr {
+                let remaining = self.wr.len();
+                trace!("writing; remaining={}", remaining);
+                let n = try_nb!(self.socket.send_to(self.wr.as_slice(), &outaddr));
+                trace!("written {}", n);
+                self.wr.clear();
+                self.out_addr = None;
+                if n != remaining {
+                    return Err(io::Error::new(io::ErrorKind::WriteZero,
+                                              "failed to write frame datagram to socket"));
+                }
+            }
+            else {
+                return Err(io::Error::new(io::ErrorKind::Other,
+                                          "outbound stream in invalid state: out_addr is not known"));
+            }
+        }
+
+        return Ok(Async::Ready(()));
+    }
+}
+
+/// Helper function that Creates a new FramedUdp object.  It moves the supplied socket, codec
+/// into the resulting FramedUdp
+pub fn framed_udp<C>(socket : UdpSocket, codec : C) -> FramedUdp<C> {
+    FramedUdp::new(
+        socket,
+        codec,
+        vec![0; 64 * 1024],
+        Vec::with_capacity(64 * 1024)
+    )
+}
+
+impl<C> FramedUdp<C> {
+
+    /// Creates a new FramedUdp object.  It moves the supplied socket, codec
+    /// supplied vecs. 
+    pub fn new(sock : UdpSocket, codec : C, rd_buf : Vec<u8>, wr_buf : Vec<u8>) -> FramedUdp<C> {
```

Comment:
I wasn't sure about this one,  since   UdpSocket::framed requires a circular dependency betwen  net/udp.rs  and io/udp_framed.rs.

I was taught that when someone asks you to create circular dependencies between API layers, you should run away screaming loudly, and tell somebody you trust.